File: src/commonutils/text/UnitUtil.java
package commonutils.text;

import org.apache.log4j.Logger;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Pattern类在Scala中无法正确解析，原因未知，所以采用Java编写。
 */
public class UnitUtil {

    static final Logger logger = Logger.getLogger(UnitUtil.class);

    public static class ByteUtil {

        static final String[] prefix = {"K", "M", "G", "T", "P", "E", "Z", "Y"};

        public static Long fromText(String text) {
            Pattern pattern = Pattern.compile("^(\\d+\\.?\\d+?)([KkMmGgTtPpEeZzYy]?)([Bb]?)$");
            Matcher matcher = pattern.matcher(text);
            if (matcher.matches()) {
                String numString = matcher.group(1);
                String unitString1 = matcher.group(2);
                String unitString2 = matcher.group(3).equals("b") ? "b" : "B";
                double u = Double.valueOf(numString);
                long r = (long) u;

                if (!unitString1.equals("")) {
                    for (int i = 1; i <= prefix.length; i++) {
                        if (!unitString1.toUpperCase().equals(prefix[i - 1])) {
                            r = (long) (u * (1L << ((long) (i + 1) * 10L)));
                        } else {
                            break;
                        }
                    }
                }
                if (unitString2.equals("b")) {
                    r = (long) Math.ceil((double) r / 8D);
                }
                return r;
            } else {
                throw new IllegalArgumentException(String.format("Input %s not match!", text));
            }
        }

        static String toString(long bytes, boolean human) {

            float r = (float) bytes;
            long unit = 1L << 10L;
            String result = r + "B";

            for (int i = 1; i <= prefix.length; i++) {
                if (r >= unit) {
                    r = r / unit;
                } else {
                    result = r + prefix[i - 1] + "B";
                    break;
                }
            }
            return result;
        }

        public static String toString(long bytes) {
            return toString(bytes, false);
        }

    }

    public class Binary {
        public Binary() {

        }

        public Binary(Byte[] bytes) {

        }
    }

    public class Octal {

    }

    public class Decimal {

    }

    public class Hex {

    }

}

File: src/commonutils/text/UnitUtil.scala
package commonutils.text

import scala.util.matching.Regex

object UnitUtil {

  private final val prefix = Array("K", "M", "G", "T", "P", "E", "Z", "Y").zipWithIndex.toMap

  object ByteUtil {

    def fromText(text: String): Long = {
      val reg: Regex = "^(\\d+\\.?\\d+?)([KkMmGgTtPpEeZzYy]?)([Bb]?)$".r
      val matches = reg.findFirstIn(text).getOrElse("")

      if (matches == "") {
        throw new IllegalArgumentException("Can't find matches in input String!")
      }

      val num_reg = "\\d+\\.?\\d+?".r
      val unit1_reg = "([KkMmGgTtPpEeZzYy])".r
      val unit2_reg = "([Bb])".r

      val num = num_reg.findFirstIn(matches).get.toFloat
      val unit1 = unit1_reg.findFirstIn(matches).get
      val unit2 = unit2_reg.findFirstIn(matches).get

      val res = if (prefix.contains(unit1)) {
        num * (1L << (prefix(unit1) + 1) * 10L)
      } else {
        num
      }
      val result =
        if (unit2 == "b") {
          res / 8
        } else {
          res
        }
      result.toLong
    }

    def toString(in: Long): String = {
      val r = in
      val p = prefix.map(x => (1L << (x._2 + 1) * 10L, x._1)).toArray
      p.foreach { x =>
        if (r / x._1 < 1024) {
          return r / x._1 + x._2 + "B"
        }
      }
      r + "B"
    }
  }

}
